File: frontend/src/lib/api/cards.ts
import { authRequest } from '$lib/stores/auth.svelte';
import { toQueryString } from './query';
import type { CardItem, CardTransactionItem, Page } from './types';

export interface CardTxnQuery {
	card_id?: number;
	approved_from?: string;
	approved_to?: string;
	merchant_category_code?: string;
	q?: string;
	include_cancelled?: boolean;
	/** 어떤 정산서에도 담기지 않은 거래만. 정산 화면의 카드내역 피커가 쓴다. */
	unsettled?: boolean;
	page?: number;
	size?: number;
}

export function listCards(): Promise<CardItem[]> {
	return authRequest<CardItem[]>('/api/v1/cards');
}

export function listCardTransactions(
	query: CardTxnQuery = {}
): Promise<Page<CardTransactionItem>> {
	return authRequest<Page<CardTransactionItem>>(
		`/api/v1/card-transactions${toQueryString(query as Record<string, unknown>)}`
	);
}

File: backend/app/routers/cards.py
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Query

from app.deps import CurrentUser, DbSession
from app.schemas.card import CardOut, CardTransactionOut
from app.schemas.common import Page
from app.services import cards as card_service

router = APIRouter(prefix="/api/v1", tags=["cards"])


@router.get("/cards", response_model=list[CardOut])
async def list_cards(user: CurrentUser, session: DbSession) -> list[CardOut]:
    return await card_service.list_my_cards(session, user=user)


@router.get("/card-transactions", response_model=Page[CardTransactionOut])
async def list_card_transactions(
    user: CurrentUser,
    session: DbSession,
    card_id: int | None = None,
    approved_from: date | None = None,
    approved_to: date | None = None,
    merchant_category_code: str | None = None,
    q: str | None = None,
    include_cancelled: bool = False,
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=100)] = 20,
) -> Page[CardTransactionOut]:
    return await card_service.list_card_transactions(
        session,
        user=user,
        filters=card_service.CardTxnFilters(
            card_id=card_id,
            approved_from=approved_from,
            approved_to=approved_to,
            merchant_category_code=merchant_category_code,
            q=q,
            include_cancelled=include_cancelled,
            page=page,
            size=size,
        ),
    )
